packages/provider/src/multi-wallet.ts

```diff
@@ -0,0 +1,271 @@
+
+    // Generate wallets using the initial configuration
+    this._wallets = params.networks.map((network) => {
+      let wallet = new Wallet(params.initialConfig, params.context, ...params.signers)
+      wallet.setProvider(network.rpcUrl)
+      if (network.relayer) {
+        wallet.setRelayer(network.relayer)
+      }
+      return {
+        network: network,
+        wallet: wallet
+      }
+    })
+
+    this.provider = this._wallets[0].wallet.provider
+  }
+
+  get address(): string {
+    return this._wallets[0].wallet.address
+  }
+
+  getAddress(): Promise<string> {
+    return this._wallets[0].wallet.getAddress()
+  }
+
+  async getFullConfig(): Promise<FullConfig> {
+    const allConfigs = await Promise.all(this._wallets.map(async (w) => ({ wallet: w, config: await this.currentConfig(w.wallet) })))
+    const thresholds = allConfigs.map((c) => ({ chaind: c.wallet.network.chainId, weight: c.config.threshold }))
+    const allSigerns = allConfigs.reduce((p, config) => {
+      config.config.signers.forEach((signer) => {
+        const item = p.find((c) => c.address === signer.address)
+        const netEntry = {
+          weight: signer.weight,
+          chaind: config.wallet.network.chainId
+        }
+
+        if (!item) {
+          p.push({
+            address: signer.address,
+            networks: [netEntry]
+          })
+        } else {
+          item.networks.push(netEntry)
+        }
+      })
+      return p
+    }, [] as Signer[])
+
+    return {
+      threshold: thresholds,
+      signers: allSigerns
+    }
+  }
+
+  async signMessage(message: Arrayish, network?: NetworkConfig): Promise<string> {
+    const wallet = network ? this.networkWallet(network) : this.mainWallet()
+
+    // TODO: Skip this step if wallet is authWallet
+    const thisConfig = await this.currentConfig(wallet)
+
+    // See if wallet has enough signer power
+    const weight = await wallet.useConfig(thisConfig).signWeight()
+    if (weight.lt(thisConfig.threshold)) {
+      throw new NotEnoughSigners(`Sign message - wallet combined weight ${weight.toString()} below required ${thisConfig.threshold.toString()}`)
+    }
+
+    return wallet.useConfig(thisConfig).signMessage(message)
+  }
+
+  async sendTransaction(transaction: Transactionish, network?: NetworkConfig | BigNumberish): Promise<TransactionResponse> {
+    const wallet = network ? this.networkWallet(network) : this.mainWallet()
+
+    // TODO: Skip this step if wallet is authWallet
+    const [thisConfig, lastConfig] = await Promise.all([
+      this.currentConfig(wallet), this.currentConfig()
+    ])
+
+    // See if wallet has enough signer power
+    const weight = await wallet.useConfig(thisConfig).signWeight()
+    if (weight.lt(thisConfig.threshold)) {
+      throw new NotEnoughSigners(`Send transaction - wallet combined weight ${weight.toString()} below required ${thisConfig.threshold.toString()}`)
+    }
```

Comment:
Not sure I understand this `(weight.lt(thisConfig.threshold))` section, would you mind explaining?

Reply:
`weight` is the combined `weight` of all signers currently configured on the wallet, the `threshold` is the `threshold` of the current configuration. If the wallet doesn't have enough signers to sign a message/transaction then this method throws an error.

Reply:
Under what condition would a wallet not have enough signers? Would have to be an invalid config, no?

Reply:
This throws when the wallet instance has not enough accessible signers (private keys + remote signers), imagine a 2/3 wallet with signers A, B, and C. If you create an instance of `arcadeum.MultiWallet` with only A private key then you need either B or C to sign transactions.

One could argues that those signers exists somewhere else and the relayer is going to aggregate the signatures later, for that maybe we can add a flag to the method `revertOnCantSign` or similar

Reply:
Yeah I guess I always assumed we would go with this assumption:

> One could argues that those signers exists somewhere else and the relayer is going to aggregate the signatures later,

I.e. that local wallets won't have many keys on their own, unless in some exceptions.

Reply:
We can always include non-local signers in the list of accessible signers of `arcadeum.Wallet` and because the `signTransaction` and `signMessage` are both async we can request the signatures of remote signers in a transparent way. This would only work when the client knows who the remote signers is, and that's the case for the Arcadeum wallet and the guard signer.